### PR TITLE
fix(validation): count bare dir refs ending at a word boundary (#3482)

### LINKED ISSUE #3482: Portability validator misses bare directory references that end at a word boundary
## Summary

`count_upstream_refs` in `scripts/validation/check_skill_md_portability.py` only
counts an upstream path when the directory name is followed by a separator, a
quote, `?`, `#`, or end of line. A reference that ends at a word boundary, a
sentence period, or a closing paren is missed.

## Evidence

Measured against both the current head of `fix/portability-boundary-followup`
and the base commit `303f918ea`, so this is not a regression from #3471:

| text | base | head | expected |
| --- | --- | --- | --- |
| `Use templates/agents for generation.` | 0 | 0 | 1 |
| ``Use `.agents` for state.`` | 0 | 0 | 1 |
| `[state](/.agents)` | 0 | 0 | 1 |
| `state` | 0 | 0 | 1 |
| `See .claude/lib in the plugin.` | 0 | 0 | 1 |

Reproduce:

```bash
uv run python -c "
import importlib.util as u
s=u.spec_from_file_location('m','scripts/validation/check_skill_md_portability.py')
m=u.module_from_spec(s); s.loader.exec_module(m)
print(m.count_upstream_refs('Use templates/agents for generation.'))"
```

## Cause

Each entry in `UPSTREAM_PATTERNS` terminates with `(?:[\\/]+|['\"?#]|$)`. That
class was written for paths that continue into a filename. A bare directory
mention ends at whitespace, `.`, `)` or `,` instead, none of which the class
accepts.

## Why it was not fixed in #3472

That PR fixes the path-start anchor. Widening the terminator changes which
references count across every skill file and would move the drift baseline, so
it needs its own change with its own before-and-after count over the whole
skill tree.

## Acceptance criteria

- The terminator accepts a word boundary as well as the current characters.
- A partial-word match such as `templates/agentsx` still does not count.
- The repository-wide count is measured before and after; any baseline movement
  is stated in the PR body.
- Positive, negative and edge cases are added to
  `tests/validation/test_check_skill_md_portability.py`.

### PR DESCRIPTION
**Pull request by rjmurillo**
## Summary

Widen the Markdown portability terminator so a bare directory reference that
ends at a word boundary is counted. Issue #3482 reported five real shapes that
the old terminator missed. The gate scanned prose for upstream-only path
prefixes but ended a reference only at a path separator, a quote, `?`, `#` or
end of string, so a name that ended at a space, a period, a comma or a closing
bracket was invisible.

Fixes #3482

## Basing decision

Based on `main`, not on #3567 (the AST parser rewrite). The terminator delta is
+36 with or without the AST walk in place (see the interaction check below), so
the change is independent of #3567 and separately reviewable. It can merge in
either order.

## What changed

Extract the terminator into one named constant and widen it. The constant
replaces five inline copies, so the anchor and terminator now have a single
source of truth.

```text
old:  (?:[\\/]+|['"?#]|$)
new:  (?:[\\/]+|(?![\w-])(?!\.[\w]))
```

`(?![\w-])` is the word-boundary case. It admits every non-identifier boundary
at once (space, tab, newline, end of string, `)`, `,`, `;`, `!`, backtick, and
the rest) and subsumes the old quote / `?` / `#` / `$` alternatives, which were
all non-identifier boundaries already. It still rejects a longer directory name,
because a trailing letter, digit, underscore or hyphen is an identifier
character.

`(?!\.[\w])` rejects a file-extension dot. Without it the case-insensitive match
counts the real file templates/AGENTS.md as the `templates/agents` directory,
because `.` is a boundary and the name folds. A sentence-ending period still
terminates, because `name.` puts a non-word character after the dot; only
`name.ext` is excluded.

## Measured baseline move

Command (runs the module's own boundary and strip over the corpus, applying the
old and new terminators side by side):

```text
uv run python equiv-scratch/term_measure.py <repo_root>
```

```text
GATE SCOPE (.claude/skills, marker-aware, the gate's real scope)
  old terminator: 0
  new terminator: 0
  delta:          0        gate stays green, no baseline update

ALL REPO raw count_upstream_refs over 3930 .md files
  old terminator: 9283
  new terminator: 9319
  delta:          +36      every one a true positive
```

The +36 breaks down as 15 `templates/agents`, 13 `.agents`, 6 `.claude/lib`
(one root-anchored), 1 `templates/platforms`, 1 `.claude/review-axes`. A naive
widening without the extension-dot guard reports +61; the extra 25 are all
templates/AGENTS.md file references colliding with the directory pattern under
the case-insensitive match, which the guard removes.

## Interaction check with #3567

The terminator delta is attributable to the terminator alone. Measured on both
bases:

```text
base = main (line-scanner strip)   old 9283  new 9319  delta +36
base = #3567 (AST-walk strip)      old 9285  new 9321  delta +36
```

The delta is +36 on both. The absolute totals differ by +2, which is exactly the
documented #3567 effect (the AST walk nets two more refs than the line scanner),
and is independent of the terminator. There is no cross-term; the two changes
compose additively.

## Terminator probe table

Every shape reported by the issue plus the extra shapes requested in review.
EXPECTED is the count the widened terminator must produce; ACTUAL is measured.

```text
NEW POSITIVES (old = 0, new = 1)                 EXPECTED  ACTUAL
  word boundary (space)                          1         1
  sentence period then space                     1         1
  sentence period then end of file               1         1
  closing paren )                                1         1
  comma ,                                        1         1
  backtick                                       1         1
  colon :                                        1         1
  semicolon ;                                    1         1
  exclamation !                                  1         1
  closing bracket ]                              1         1
  closing brace } (space anchored)               1         1
  end of line, interior                          1         1
  non-ASCII non-letter (arrow)                   1         1

REGRESSION GUARDS (old = 1, new must stay 1)     EXPECTED  ACTUAL
  end of file, no trailing newline ($)           1         1
  possessive apostrophe                          1         1
  double quote                                   1         1
  question mark                                  1         1
  fragment hash #section                         1         1

NEGATIVES (old = 0, new = 0)                      EXPECTED  ACTUAL
  followed by letter (agentsx)                   0         0
  followed by digit (agents2)                    0         0
  followed by underscore (agents_v2)             0         0
  followed by hyphen (agents-v2)                 0         0
  file extension (agents.md)                     0         0
  case collision (AGENTS.md)                     0         0
  non-ASCII letter (different name)              0         0
  brace not an anchor ({templates/agents})       0         0
```

The brace probe needs a valid anchor: `{` is deliberately not an anchor (issue
#3489 rejects template-variable shapes), so a leading `{` yields 0, while a
space-anchored `templates/agents}` counts because the `}` is a terminator.

## Revert proof, both directions

Command:

```text
uv run python equiv-scratch/revert_proof.py <repo_root>
```

Each positive returns 0 under the old terminator, so its test fails on revert.
Each negative returns 0 under the new terminator and 1 under a plausible wrong
widening, so it is non-vacuous. Each regression guard returns 1 under both.

```text
POSITIVES        old  new  naive  bare-\b     these fail if the widening reverts
  all 13 shapes  0    1    1      1

NEGATIVES        old  new  naive  bare-\b     guard catches
  agentsx        0    0    0      0           acceptance criterion
  agents2        0    0    0      0           acceptance criterion
  agents_v2      0    0    0      0           acceptance criterion
  agents-v2      0    0    0      1           bare-\b widening
  agents.md      0    0    1      1           naive no-dot-guard widening
  AGENTS.md      0    0    1      1           naive no-dot-guard widening
  non-ASCII lett 0    0    0      0           acceptance criterion

REGRESSION GUARD old  new
  eof no newline 1    1
  possessive     1    1
  double quote   1    1
  question mark  1    1
```

Suite-level RED/GREEN confirms the same: reverting the terminator constant to its
old value makes 14 tests fail (13 new positives plus the drift CLI test) and
leaves 15 passing (5 old-boundary guards, 9 negatives, and the clean CLI test),
proving the negatives and guards do not depend on the widening.

## Type of change

Bug fix. Behavior change to the ratchet's counting, no interface change.

## Testing

```text
RED   (terminator reverted): 14 failed, 15 passed in the new class
GREEN (terminator widened):  29 passed in the new class
FILE  (whole test module):   148 passed, 0 failed
```

Positive, negative and edge cases with asserted CLI exit codes (drift returns 1,
clean returns 0). Mocked filesystem via `tmp_path`. Gate scope stays 0 to 0, so
no baseline JSON changes and the committed-repo drift test still passes.

## References

Reference-only paths, not part of this diff:

```text
scripts/utils/markdown_parser.py
templates/agents/security.shared.md
templates/platforms/copilot-cli.yaml
.agents/governance/TESTING-RIGOR.md
```

### CODE REVIEW

**Comment by rjmurillo** [comment]
owner: local:pr-autofix
session: 2026-07-28-session-3608
acquired_at: 2026-07-28T08:10:23Z
expires_at: 2026-07-28T08:25:23Z
base_sha: 87b1176674a8f444a5391a34e9cac187d7af024f

File: scripts/validation/check_skill_md_portability.py [hunk 1]
```diff
@@ -150,19 +150,48 @@
...
+# consumes the separators, so ``/templates/agents/x.md`` counts and an adjacent
+# reference is not double counted.
+#
+# ``(?![\w-])`` is the word-boundary case the older class missed. A bare mention
+# such as ``templates/agents for generation`` ends at a space, and
+# ``[state](/.agents)`` ends at ``)``; both are boundaries, neither a separator
+# nor one of the few punctuation marks the old class named. The negative
+# lookahead admits every non-identifier boundary at once (space, tab, newline,
+# end of string, ``)``, ``,``, ``;``, ``!``, backtick, ``'``, ``"``, ``?``,
+# ``#`` and the rest) while still rejecting a longer directory name:
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/validation/check_skill_md_portability.py]: Fixed in 60e8bebc: the punctuation list is now spelled out in prose (closing parenthesis, comma, semicolon, and so on) instead of nested code spans, removing the malformed comma literal.
